storm-core/src/jvm/org/apache/storm/daemon/nimbus/Nimbus.java

```diff
@@ -0,0 +1,3729 @@
+                    LOG.info("desired replication count of {} not achieved but we have hit the max wait time {}"
+                            + " so moving on with replication count for conf key = {} for code key = {} for jar key = ",
+                            minReplicationCount, maxWaitTime, confCount, codeCount, jarCount);
+                    return;
+                }
+                LOG.info("WAITING... {} <? {} {} {}", minReplicationCount, jarCount, codeCount, confCount);
+                LOG.info("WAITING... {} <? {}", totalWaitTime, maxWaitTime);
+                Time.sleepSecs(1);
+                totalWaitTime++;
+                if (!ConfigUtils.isLocalMode(topoConf)) {
+                    jarCount = getBlobReplicationCount(ConfigUtils.masterStormJarKey(topoId));
+                }
+                codeCount = getBlobReplicationCount(ConfigUtils.masterStormCodeKey(topoId));
+                confCount = getBlobReplicationCount(ConfigUtils.masterStormConfKey(topoId));
+            }
+        }
+        LOG.info("desired replication count {} achieved, current-replication-count for conf key = {},"
+                + " current-replication-count for code key = {}, current-replication-count for jar key = {}", 
+                minReplicationCount, confCount, codeCount, jarCount);
+    }
+    
+    private TopologyDetails readTopologyDetails(String topoId) throws NotAliveException, KeyNotFoundException, AuthorizationException, IOException, InvalidTopologyException {
+        StormBase base = stormClusterState.stormBase(topoId, null);
+        if (base == null) {
+            if (topoId == null) {
+                throw new NullPointerException();
+            }
+            throw new NotAliveException(topoId);
+        }
+        BlobStore store = blobStore;
+        Map<String, Object> topoConf = readTopoConfAsNimbus(topoId, store);
+        StormTopology topo = readStormTopologyAsNimbus(topoId, store);
+        Map<List<Integer>, String> rawExecToComponent = computeExecutorToComponent(topoId);
+        Map<ExecutorDetails, String> executorsToComponent = new HashMap<>();
+        for (Entry<List<Integer>, String> entry: rawExecToComponent.entrySet()) {
+            List<Integer> execs = entry.getKey();
+            ExecutorDetails execDetails = new ExecutorDetails(execs.get(0), execs.get(1));
+            executorsToComponent.put(execDetails, entry.getValue());
+        }
+        
+        return new TopologyDetails(topoId, topoConf, topo, base.get_num_workers(), executorsToComponent, base.get_launch_time_secs());
+    }
+    
+    private void updateHeartbeats(String topoId, Set<List<Integer>> allExecutors, Assignment existingAssignment) {
+        LOG.debug("Updating heartbeats for {} {}", topoId, allExecutors);
+        IStormClusterState state = stormClusterState;
+        Map<List<Integer>, Map<String, Object>> executorBeats = StatsUtil.convertExecutorBeats(state.executorBeats(topoId, exis
```

Comment:
If we want to fix an issue due to STORM-2126 I would propose that we do that on a separate JIRA.  Mostly because STORM-2126 went into the 1.1 line as well as the 2.0 line.  If that goes in first I would then be happy to translate it to java.